### ISSUE TITLE
[css-device-adapt] Remove @viewport

### ISSUE THREAD

**Issue by smfr** [w3c/csswg-drafts#4766]
https://www.w3.org/TR/css-device-adapt-1/#atviewport-rule

The @viewport rule in this spec has not gained traction. I propose that we remove it, and use the spec to simply define the `<meta name="viewport">` tag.

**Comment by rolivo** [comment]
Why is the solution always to eliminate, destroy, restrict ... why not leave it for the next level of the specification?

**Comment by smfr** [comment]
Issue #258 has some relevant discussion. If browsers are unwilling to implement, then there's no point keeping @viewport around. Punting it to the next level just kicks the can down the road.

**Comment by frivoal** [comment]
I like `@viewport`, but I cannot disagree about its lack of traction. In fact, it's lost traction: it used to ship in Opera and IE, but with the demise of those as standalone engines, it's no longer in any modern browser. That makes me sad, but it's nonetheless true.

Also, the viewport meta needs a spec.

Also, we recently discussed specifying the definitions of layout viewport and visual viewport, and temporarily agreed that they could go in cssom-view, but there was some reluctance about that due to the fact that not all implementations that care about the viewport have an OM, so the coupling felt wrong.

With all that in mind, moving the definitions of the viewports here, defining the viewport meta alongside, and removing the rest seems sensible to me.

There are a number of useful things that `@viewport` does/did that the meta viewport doesn't do, so it would probably be good to set up an explainer somewhere to preserve those, in the hope that we'll get back to this some day, but the current state of affairs doesn't help anyone.

**Comment by TalbotG** [comment]
> * RESOLVED: Remove @viewport, retire the spec, 
> move the meta tag to a new spec (CSS Viewport)

There are 2 tests specifically testing @viewport :
http://wpt.live/css/css-values/vh-support-atviewport.html
http://wpt.live/css/css-device-adapt/idlharness.html
Should we remove such tests?

**Comment by smfr** [comment]
Yes.

**Comment by karlcow** [comment]
@RByers is there an intent to unship this?
https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/css/resolver/viewport_style_resolver.cc?l=85-113&rcl=7e6122a30da5d3a0ffded6610e526934c0fa21b7

**Comment by astearns** [comment]
My assumptions on what we need to do with publishing and process for all of this

- Create an ED of css-viewport
- Get a resolution on FWPD of css-viewport
- Get a resolution on publishing css-device-adaptation as a NOTE

I think it would be best if the publishing all happened at once, with something in the NOTE version of css-device-adaptation pointing to the new module.

**Comment by ddbeck** [comment]
Hi folks! Do you still plan to remove `@viewport` from the spec? I'd like to update the [compatibility data on MDN](https://github.com/mdn/browser-compat-data/issues/6213) to reflect the state of this feature. Having it definitively removed or knowing that it's unlikely to happen soon would be helpful for that. Thank you!

**Comment by Malvoz** [comment]
@ddbeck per https://github.com/w3c/csswg-drafts/issues/4766#issuecomment-585314599:

> `RESOLVED: Remove @viewport, retire the spec, move the meta tag to a new spec (CSS Viewport)`

**Comment by karlcow** [comment]
> My assumptions on what we need to do with publishing and process for all of this
> 
>     * Create an ED of css-viewport

@astearns do you know if it has been started?

**Comment by astearns** [comment]
I expect not. @smfr, am I wrong?

**Comment by Zhang-Junzhi** [comment]
> There are a number of useful things that `@viewport` does/did that the meta viewport doesn't do, so it would probably be good to set up an explainer somewhere to preserve those, in the hope that we'll get back to this some day, but the current state of affairs doesn't help anyone.

I also hope so. meta viewport doesn't. Letting HTML tags care about layout doesn't meet the fundamentalthe principle of HTML. This should return to CSS someday.

Is there any plan to make a similar feature come back to CSS in the near future?

**Comment by IllidanS4** [comment]
I have a number of pages, some auto-generated and with limited control of their actual layout or the tags, but I have control of the CSS that is applied to them all. A `meta` tag is the *worst* place where something that seems to fix displaying these pages on a phone should be placed. It is the style that controls whether a page is optimized for phones, not the content.

**Comment by kevinwr** [comment]
Not sure if this is the best place to ask, but I figured I'd give it a go. What got me here initially was wondering how the  element `<meta name="viewport" content="width=device-width; initial-scale=1.0">` was being processed. That lead me to the [CSS Device Adaptation](https://drafts.csswg.org/css-device-adapt/) page.

It is my understanding that the browser then parses this line and translates it into an `@viewport` rule and then inserts it into the DOM as if it were in a `<style>` element. I was then shocked to find out this CSS rule is deprecated, leading to another question and why I'm here. My question is if the `@viewport` rule is deprecated, then how can the `<meta name="viewport" content="width=device-width; initial-scale=1.0">` HTML tag still work?

Keep in mind, I'm extremely new to the field of web development and so I don't yet understand how things happen "under-the-hood" so to speak. Please be gentle. Thank you.

**Comment by karlcow** [comment]
@kevinwr On Firefox side some of your questions are here.

https://searchfox.org/mozilla-central/rev/6b099d836c882bc155d2ef285e0ad0ab9f5038f6/dom/base/nsViewportInfo.h
https://searchfox.org/mozilla-central/rev/6b099d836c882bc155d2ef285e0ad0ab9f5038f6/dom/base/Document.cpp#9909

@smfr is the person who may have the start of a draft but it's not public yet for how it is working. 
https://github.com/w3c/csswg-drafts/issues/4766#issuecomment-642148468

**Comment by jmi2k** [comment]
I agree with @Zhang-Junzhi and @IllidanS4. The `meta` tag is an style element, it should be inside CSS somehow. Having it in HTML is an ugly hack, and removing it from CSS just because browsers don't want to implement sounds like a fairly weak argument. Standards are there precisely to guide implementations towards correct, sound solutions.

tl;dr: get rid of that `meta` tag for setting up the viewport, gives us a CSS solution instead

**Comment by ByteEater-pl** [comment]
I'm not sure if [the comment I made](https://github.com/w3c/csswg-drafts/issues/331#issuecomment-955610799) on #331 would be better placed here. Anyway, linking to let people discussing and reading here know.
